File: Perl/132 Large repunit factors -- v2.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# Date: 21 September 2019
# https://github.com/trizen

# https://projecteuler.net/problem=132

# Runtime: 0.051s

# See also:
#   https://oeis.org/A178070

use 5.010;
use strict;
use warnings;

use ntheory qw(:all);

my $n     = 1e9;
my $count = 40;

my $factors_sum = 0;

for (my $p = 7 ; ; $p = next_prime($p)) {

    my $z = znorder(10, $p);

    if ($n % $z == 0) {
        $factors_sum += $p;
        last if --$count == 0;
    }
}

say $factors_sum;

File: Julia/132 Large repunit factors -- v2.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# Date: 16 November 2021
# https://github.com/trizen

# https://projecteuler.net/problem=132

# Runtime: 1.133s

using Primes

function divisors(n)

    d = Int64[1]

    for (p,e) in factor(n)
        t = Int64[]
        r = 1

        for i in 1:e
            r *= p
            for u in d
                push!(t, u*r)
            end
        end

        append!(d, t)
    end

    return sort(d)
end

function prime_znorder(a, n)
    for d in divisors(n-1)
        if (powermod(a, d, n) == 1)
            return d
        end
    end
end

function p_132()

    N = 1e9
    C = 40

    count = 0
    total = 0

    p = 7
    while true
        if (rem(N, prime_znorder(10, p)) == 0)
            total += p
            count += 1
            count >= C && break
        end
        p = nextprime(p+1)
    end

    return total
end

println(p_132())
